### PR TITLE
Clean up HTTP client

### PR DESCRIPTION
**Pull request by jerith**

### CODE REVIEW

**Comment by jerith** [comment]
I plan to fix the race condition in the streaming client (by waiting for the async response event, at least) and maybe do a few more cleanup type things before this is ready to merge, but it's ready for an initial review at least.

File: lib/marathon_client.ex
Comment (JayH5) [inline lib/marathon_client.ex]: I'm confused about which side of the `=` what is... 🤔

File: test/marathon_client_test.exs
Comment (JayH5) [inline test/marathon_client_test.exs]: `65` seems like an odd number?

File: test/test_helper.exs
Comment (JayH5) [inline test/test_helper.exs:46]: Is there any way to define a protocol for this kind of thing?

**Review by JayH5 — COMMENTED** [review]
Seems reasonable so far.

File: lib/marathon_client.ex
Comment (jerith) [reply lib/marathon_client.ex]: It's a pattern match, and I believe you can have as many `=` as you like:
```
iex(1)> a = {b, c} = {3, 5}; {a, b, c}
{{3, 5}, 3, 5}
```

The `msg=%HTTPoison.AsyncChunk{chunk: chunk}` is matching the arg value against both `msg` (which captures the whole struct) and `%HTTPoison.AsyncChunk{chunk: chunk}` (which extracts just the chunk) so that the body of the function can use both `msg` and `chunk`.

File: test/marathon_client_test.exs
Comment (jerith) [reply test/marathon_client_test.exs]: Of course. It would have to be divisible by two to be even... :trollface: 

I carefully calculated the optimal value to use by asking the person next to me "quick, give me a number between ten and a hundred" and then checking that their response made the tests pass consistently about a dozen times.

File: test/test_helper.exs
Comment (jerith) [reply test/test_helper.exs:46]: This is a loop handler, described at https://ninenines.eu/docs/en/cowboy/2.1/guide/loop_handlers/ in the cowboy docs. The cowboy request handler process calls `init/2` (as with a normal handler) and then, if `init/2` returns a `{:cowboy_loop, ...}` tuple, waits for messages and passes them to `info/3` for further processing.

In Elixir, "protocol" has a specific meaning that doesn't seem to apply here. Did you means something else?

File: lib/marathon_client.ex
Comment (JayH5) [reply lib/marathon_client.ex]: So... `%HTTPoison.AsyncChunk{chunk: chunk}=msg` would do the same thing?

Comment (jerith) [reply lib/marathon_client.ex]: Yes.

File: test/test_helper.exs
Comment (JayH5) [reply test/test_helper.exs:46]: I knew it had a specific meaning, but guess I meant in the general sense. I think I actually misunderstood and thought that anything that could listen for SSE events needed to have a `close` but this is just a `FakeMarathon` thing for tests.

Comment (jerith) [reply test/test_helper.exs:46]: Ah. I'm not aware of any mechanism for specifying the set of messages a particular process will accept, but the standard way to do this sort of thing is with functions that send the appropriate messages. I'll add those to this handler module to make things more explicit.

File: lib/marathon_client.ex
Comment (JayH5) [reply lib/marathon_client.ex]: Do we want to be consistent? In the same line you have `{_, ssep}=state` which is the other way round.

Comment (jerith) [reply lib/marathon_client.ex]: Consistency is good. I'm probably going to remove most of that debug logging soon, but I'll fix the ordering for now.

File: test/marathon_client_test.exs
Comment (jerith) [inline test/marathon_client_test.exs:92]: This log capturing is mostly to avoid an ugly red error in the test output without silencing any other error logs we might want to know about.

File: lib/marathon_client.ex
Comment (JayH5) [inline lib/marathon_client.ex:33]: I'm super cautious about timeouts after my Twisted experiences. So I'd like to be clear about what exactly this timeout means. But maybe in a later PR we can work on timeouts specifically and leave this for now.

Comment (jerith) [reply lib/marathon_client.ex:33]: I haven't looked at the implementation under the hood, but the exhibited behaviour is that that we receive an `%HTTPoison.Error{}` struct with `{:closed, :timeout}` as the reason. We handle that by returning `{:stop, reason, state}` from our `handle_info` function, which shuts down the `GenServer` and (because the reason isn't `:normal`) any linked processes. I would expect hackney to do the necessary cleanup at that point, and unlike Erlang/Elixir makes that a whole lot easier than Python+Twisted does.

I should also note that there are two timeout parameters we can specify in hackney calls. `:timeout` (which we leave at the default) is the connect timeout and `:recv_timeout` (which we set here) is the "time between response bytes arriving" that we care about for the stream.

We could also have set the hackney timeout high and used timers in the `GenServer` to have more fine-grained control, but that would be a bunch more work and isn't really necessary unless we run into some kind of problem with the hackney timeouts.

Comment (JayH5) [reply lib/marathon_client.ex:33]: Cool, I didn't really manage to find good timeout docs for HTTPoison. So I wasn't sure what exactly `recv_timeout` was.

**Review by JayH5 — APPROVED** [review]
👍

File: lib/marathon_client.ex
Comment (jerith) [reply lib/marathon_client.ex:33]: Yeah. A lot of those parameters are just passed directly through to hackney, so you need to read the hackney docs to understand them. Apart from the high-level API bits, I've mostly been working from the hackney docs for this stuff.

**Pull request merged**
